cpp/dbscan.cpp

```diff
@@ -52,8 +52,12 @@ auto Dbscan::fit_predict(std::vector<Dbscan::Point> const& points) -> std::vecto
 
     // FIRST PASS OVER THE POINTS
     for (auto const& pt : points) {
-        auto const bin_x{static_cast<std::uint32_t>(std::floor((pt[0] - min[0]) / eps_))};
-        auto const bin_y{static_cast<std::uint32_t>(std::floor((pt[1] - min[1]) / eps_))};
+        auto bin_x{static_cast<std::uint32_t>(std::floor((pt[0] - min[0]) / eps_))};
+        auto bin_y{static_cast<std::uint32_t>(std::floor((pt[1] - min[1]) / eps_))};
+
+        bin_x = bin_x < num_bins_x ? bin_x : num_bins_x - 1;
+        bin_y = bin_y < num_bins_y ? bin_y : num_bins_y - 1;
+
```

Comment:
The issue here is not so much having two input points, but that for your input data `(pt[0] - min[0]) / eps_` divides without any remainder and so `std::floor` becomes a no-op here. Concretely, we do `(1.0 - 0.0) / 0.5 == 2.0`. The `std::ceil` we do above to determine the number of bins also does nothing here, because `range_x`, which happens to be 1.0, is also divisible without remainder by 0.5.

Instead of clamping the index calculation, I'd rather change the way we determine the number of bins:

```c++
    // add 1e-8 to handle the case where range is exactly divisible by eps_
    auto const num_bins_x{static_cast<std::uint32_t>(std::ceil((range_x + 1e-8f) / eps_))};
    auto const num_bins_y{static_cast<std::uint32_t>(std::ceil((range_y + 1e-8f) / eps_))};
```